### PR TITLE
refactor & fix: re-organized the scripts in Enzeptional and added a test script

### PR DESCRIPTION
**Pull request by yvesnana**

### CODE REVIEW

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 1]
```diff
@@ -1,17 +1,34 @@
...
 from tape.datasets import pad_sequences
 from tape.registry import registry
 from tape.tokenizers import TAPETokenizer
-from transformers import AutoModelWithLMHead, AutoTokenizer
+from transformers import AutoModelForMaskedLM, AutoTokenizer, T5Model, T5Tokenizer
 
 from ..torch import device_claim
 
+
+def merge_dictionaries(lst_Dict):
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:20]: `lst_Dict` mixes snake and camel case, for variables and methods we use snake, so `list_of_dictionaries` would be best.
Also add docstring.

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 2]
```diff
@@ -116,6 +128,89 @@ def __call__(self, samples: List[str]) -> np.ndarray:
...
+
+        with torch.no_grad():
+            embedding = self.model(
+                input_ids=input_ids,
+                attention_mask=attention_mask,
+                decoder_input_ids=input_ids,
+            )
+
+        # For feature extraction we recommend to use the encoder embedding
+        encoder_embedding = embedding[2].cpu().numpy()
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:199]: could we conditionally select cpu or gpu here.

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 3]
```diff
@@ -137,16 +231,14 @@ def __init__(
         self.device = device_claim(device)
         # tokenizer and model definition
         self.tokenizer = AutoTokenizer.from_pretrained(model_name)
-        self.model = AutoModelWithLMHead.from_pretrained(tokenizer_name)
+        self.model = AutoModelForMaskedLM.from_pretrained(tokenizer_name)
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:234]: out of curiosity, why changing this here?

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 4]
```diff
@@ -232,3 +324,39 @@ def reconstruct_sequence_with_mutation_range(
         mutated_sequence_offset = end + 1
     mutated_sequence += sequence[end + 1 :]
     return mutated_sequence
+
+
+def selection(scores, k=20):
+    """Selecte the top K mutated sequences based on the score
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:330]: check the docstrings there are typos, K should be lower case and we also have to report the default.

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 5]
```diff
@@ -232,3 +324,39 @@ def reconstruct_sequence_with_mutation_range(
...
+        retuen the top K sequences
+
+    """
+
+    res = list(sorted(scores, key=lambda d: d["score"], reverse=True))[:k]
+
+    return res
+
+
+def crossover(p1, p2):
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:346]: typing missing

File: src/gt4sd/frameworks/enzeptional/processing.py (hunk 4 above)
Comment (drugilsberg) [reply src/gt4sd/frameworks/enzeptional/processing.py:330]: typing missing

**Review by drugilsberg — CHANGES_REQUESTED** [review]
Left some comments, in general we are going in the right direction but we need to refactor a bit more. Besides the comments here additional things to consider before merging:
- unit tests with pytests have to be written, on for each file where we have tests (they will replace the single `test_enzeptional.py`).
- path to files have to be replaced with `pkg_resources` and the files added to `setup.cfg` to be tracked.
- let's rename the scorer in the tests folder to `scorer.pkl`.

File: src/gt4sd/frameworks/enzeptional/genetic_algorithm.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+                    visited_sequences.add(mutated_sequence)
+                    population.append(mutated_sequence)
+
+            population_scores = []
+            for i in range(
+                0, len(population), number_of_samples_per_step
+            ):  # score the population in chunks
+
+                # evaluate all candidates in the population
+                if (
```
[the remark this answers is not in the document]
Comment (yvesnana) [reply src/gt4sd/frameworks/enzeptional/genetic_algorithm.py:150]: That's true. My only issue here is that I would like not to call/load the models multiple times

File: src/gt4sd/frameworks/enzeptional/genetic_algorithm.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+            [self.sequence[start:end] for start, end in intervals]
+        )
+
+        # mutate the sequence from intervals
+        aa_sequence_range = AASequence(sequence_from_intervals, mutations=mutations)
+        maximum_number_of_mutations = number_of_mutations
+        logger.info(
+            f"maximum number of mutations for the intervals: {maximum_number_of_mutations}"
+        )
+        if self.protein_embedding == "ProtTransXL" in str(type(self.protein_embedding)):
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/genetic_algorithm.py:107]: this need to be changed, `self.protein_embedding` should be a `StringEmbedding` class, then everything should become consistent.

File: src/gt4sd/frameworks/enzeptional/genetic_algorithm.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+                    visited_sequences.add(mutated_sequence)
+                    population.append(mutated_sequence)
+
+            population_scores = []
+            for i in range(
+                0, len(population), number_of_samples_per_step
+            ):  # score the population in chunks
+
+                # evaluate all candidates in the population
+                if "ProtTransXL" in str(type(self.protein_embedding)):
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/genetic_algorithm.py:153]: same comment as above

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 9]
```diff
@@ -116,6 +128,89 @@ def __call__(self, samples: List[str]) -> np.ndarray:
         )
 
 
+class ProtTransXL(StringEmbedding):
+    """Embed a string representation of a molecule using a ProtTrans_XLnet model"""
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:132]: are we sure the docstring and the class name are correct? to me it looks like the model is T5.

Comment (drugilsberg) [reply src/gt4sd/frameworks/enzeptional/processing.py:132]: but maybe it's a generic one

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 10]
```diff
@@ -116,6 +128,89 @@ def __call__(self, samples: List[str]) -> np.ndarray:
...
+        attention_mask = torch.tensor(token_ids["attention_mask"]).to(self.device)
+
+        with torch.no_grad():
+            embedding = self.model(
+                input_ids=input_ids,
+                attention_mask=attention_mask,
+                decoder_input_ids=input_ids,
+            )
+
+        # For feature extraction we recommend to use the encoder embedding
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:198]: comments start lower case

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 11]
```diff
@@ -116,6 +128,89 @@ def __call__(self, samples: List[str]) -> np.ndarray:
...
+        encoder_embedding = embedding[2].cpu().numpy()
+        # decoder_embedding = embedding[0].cpu().numpy()
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:200]: remove commented out code, not needed

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 12]
```diff
@@ -74,12 +90,10 @@ def __init__(
         self.model.eval()
         self.tokenizer = TAPETokenizer(vocab=aa_vocabulary)
 
-    def _encode_and_mask(self, sequence: str) -> Tuple[np.ndarray, np.ndarray]:
+    def _encode_and_mask(self, sequence: str) -> Tuple[Any, Any]:
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:93]: do we really need to use `Any`?

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 13]
```diff
@@ -232,3 +324,39 @@ def reconstruct_sequence_with_mutation_range(
         mutated_sequence_offset = end + 1
     mutated_sequence += sequence[end + 1 :]
     return mutated_sequence
+
+
+def selection(scores: List[Dict[str, Any]], k: int = 20) -> List[Any]:
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:329]: this should leave in `genetic_algorithm.py`

File: src/gt4sd/frameworks/enzeptional/processing.py [hunk 14]
```diff
@@ -232,3 +324,39 @@ def reconstruct_sequence_with_mutation_range(
...
+        retuen the top K sequences
+
+    """
+
+    res = list(sorted(scores, key=lambda d: d["score"], reverse=True))[:k]
+
+    return res
+
+
+def crossover(p1: str, p2: str) -> Tuple[str, str]:
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/processing.py:346]: this should leave in `genetic_algorithm.py`

Comment (drugilsberg) [reply src/gt4sd/frameworks/enzeptional/processing.py:346]: those `p1` and `p2` do not help as variable names, maybe `a_sequence` and `another_sequence` are better

File: src/gt4sd/frameworks/enzeptional/tests/test_enzeptional.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
+# %%
```
Comment (drugilsberg) [inline src/gt4sd/frameworks/enzeptional/tests/test_enzeptional.py:1]: the tests should be using pytest

**Review by drugilsberg — CHANGES_REQUESTED** [review]
See my comments, there is still some open points that need to be addressed.

Also, rename: `src/gt4sd/frameworks/enzeptional/tests/small_exmaple_ProtTrans_scorrer.pkl` to `src/gt4sd/frameworks/enzeptional/tests/scorer.pkl`.
And remember to add both the scorer and `mutation_scheme.json` to `setup.cfg`.

**Pull request closed without merging**